### ISSUE TITLE
Calling torch._C._cuda_attach_out_of_memory_observer crashes

### ISSUE THREAD

**Issue by alito** [pytorch/pytorch#102796]
### 🐛 Describe the bug

Trying to attach an OOM observer triggers an immediate segfault.

```
% cat attach.py
```
```python
def reporter(*args, **kwargs):
    print("in reporter")

import torch
torch._C._cuda_attach_out_of_memory_observer(reporter)
```

```
% python3 attach.py
zsh: segmentation fault (core dumped)  python3 attach.py
% gdb ~/.virtualenvs/torchedge/bin/python core.25885
...
(gdb) bt
#0  0x00007f5fcc43873b in c10::cuda::CUDACachingAllocator::Native::NativeCachingAllocator::attachOutOfMemoryObserver(std::function<void (long, long, long, long)>) ()
   from /home/dubro/.virtualenvs/torchedge/lib/python3.9/site-packages/torch/lib/libc10_cuda.so
#1  0x00007f600bf2610a in THCPModule_attachOutOfMemoryObserver(_object*, _object*) () from /home/dubro/.virtualenvs/torchedge/lib/python3.9/site-packages/torch/lib/libtorch_python.so
#2  0x000055b850a1d596 in cfunction_vectorcall_O () at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Objects/methodobject.c:512
#3  0x000055b850a679c5 in _PyObject_VectorcallTstate (kwnames=0x0, nargsf=<optimized out>, args=0x55b851accd08, callable=0x7f5faf2e54f0, tstate=0x55b851a73140)
    at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Include/cpython/abstract.h:127
#4  PyObject_Vectorcall (kwnames=0x0, nargsf=<optimized out>, args=0x55b851accd08, callable=0x7f5faf2e54f0) at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Include/cpython/abstract.h:127
#5  call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>, tstate=0x55b851a73140) at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/ceval.c:5072
#6  _PyEval_EvalFrameDefault () at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/ceval.c:3487
#7  0x000055b8509bd560 in _PyEval_EvalFrame (throwflag=0, f=0x55b851accb90, tstate=0x55b851a73140) at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Include/internal/pycore_ceval.h:40
#8  _PyEval_EvalCode () at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/ceval.c:4327
#9  0x000055b850aa2227 in _PyEval_EvalCodeWithName (_co=<optimized out>, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=<optimized out>, kwnames=<optimized out>, kwargs=0x0, kwcount=0, kwstep=2,
    defs=0x0, defcount=0, kwdefs=0x0, closure=0x0, name=0x0, qualname=0x0) at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/ceval.c:4359
#10 0x000055b850aa2269 in PyEval_EvalCodeEx () at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/ceval.c:4375
#11 0x000055b850aa228b in PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>) at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/ceval.c:826
#12 0x000055b850ad4dc9 in run_eval_code_obj () at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/pythonrun.c:1219
#13 0x000055b850b0f894 in run_mod () at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/pythonrun.c:1240
#14 0x000055b850998e4a in pyrun_file (fp=0x55b851ad4ec0, filename=0x7f604a3d37b0, start=<optimized out>, globals=0x7f604a402b40, locals=0x7f604a402b40, closeit=1, flags=0x7ffce1dee878)
    at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/pythonrun.c:1138
#15 0x000055b850b15a02 in pyrun_simple_file (flags=0x7ffce1dee878, closeit=1, filename=0x7f604a3d37b0, fp=0x55b851ad4ec0) at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/pythonrun.c:449
#16 PyRun_SimpleFileExFlags () at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Python/pythonrun.c:482
#17 0x000055b850b160d5 in pymain_run_file (cf=0x7ffce1dee878, config=0x55b851a71cc0) at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Modules/main.c:379
#18 pymain_run_python (exitcode=0x7ffce1dee870) at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Modules/main.c:604
#19 Py_RunMain () at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Modules/main.c:683
#20 0x000055b850b16229 in Py_BytesMain (argc=<optimized out>, argv=<optimized out>) at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Modules/main.c:1129
#21 0x00007f60493d013a in __libc_start_main () from /lib64/libc.so.6
#22 0x000055b850a8f3a1 in _start () at /home/conda/feedstock_root/build_artifacts/python-split_1620683970095/work/Parser/pegen/pegen.c:55
```

### Versions

```
% python3 collect_env.py
Collecting environment information...
PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Amazon Linux 2 (x86_64)
GCC version: (GCC) 7.3.1 20180712 (Red Hat 7.3.1-15)
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.26

Python version: 3.9.4 | packaged by conda-forge | (default, May 10 2021, 22:13:33)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-4.14.290-217.505.amzn2.x86_64-x86_64-with-glibc2.26
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 510.73.08
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  2
Core(s) per socket:  2
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Platinum 8259CL CPU @ 2.50GHz
Stepping:            7
CPU MHz:             3099.998
BogoMIPS:            5000.00
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            36608K
NUMA node0 CPU(s):   0-3
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke avx512_vnni

Versions of relevant libraries:
[pip3] torch==2.0.1
[pip3] triton==2.0.0
[conda] No relevant packages
```

**Comment by awgu** [comment]
I am not sure if there is a label for this, so just directly cc-ing @zdevito.

**Comment by alito** [comment]
Calling `torch.cuda.memory._record_memory_history` before registering the OOM observer stops it from crashing. Strange, but this is all internal, so can't expect it to be documented. Closing.

**Comment by zdevito** [comment]
I think this happens because there is a missing check to make sure the allocator is initialized before the call is made. We can add the check or put it in a wrapper function that initializers the allocator first.
